product.go

```diff
@@ -232,6 +237,45 @@ func (s *ProductServiceOp) Delete(productID int64) error {
 	return s.client.Delete(fmt.Sprintf("%s/%d.json", productsBasePath, productID))
 }
 
+func (s *ProductServiceOp) IterateAll(f func(product Product) error) error {
+	return s.Iterate(ProductListOptions{}, f)
+}
+
+// iterate through products one at a time and passes them to the helper func
+// uses since_id and order will be ignored, use ListWithPagination if you care about order
+func (s *ProductServiceOp) Iterate(options ProductListOptions, f func(product Product) error) error {
+	if options.Limit <= 0 {
+		options.Limit = productsMaxLimit
+	}
+
+	sinceId := int64(0)
+	for {
+		if sinceId > int64(0) {
+			options.SinceID = sinceId
+		}
+
+		products, err := s.List(options)
+
+		if err != nil {
+			return err
+		}
+
+		if len(products) == 0 {
+			break
+		}
+
+		for _, product := range products {
+			if err := f(product); err != nil {
+				return err
+			}
+
+			sinceId = product.ID
+		}
```

Comment:
adding the check ```if len(product)<options.Limit { break }```  here will allow to save one call to Shopify.
I.e. if we get less records than we requested - there is no next page.